### PR TITLE
feat: support cte in dphyp

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. pruning search space
2. make cte run in dphyp algo
3. let subquery run in new dphyp
4. make tpcds64 run!!!

Closes #issue

### CODE REVIEW

**Comment by xudong963** [comment]
@mergify update

**Comment by Xuanwo** [comment]
> 4\. make tpcds64 run!!!

:rocket: :rocket: :rocket:

File: src/query/sql/src/planner/optimizer/rule/rewrite/rule_eliminate_eval_scalar.rs [hunk 1]
```diff
@@ -65,24 +63,23 @@ impl Rule for RuleEliminateEvalScalar {
...
+        // .collect::<Vec<ScalarItem>>();
+        // if new_items.is_empty() {
+        // state.add_result(s_expr.child(0)?.clone());
+        // } else {
+        // state.add_result(SExpr::create_unary(
+        // EvalScalar { items: new_items }.into(),
+        // s_expr.child(0)?.clone(),
+        // ));
+        // }
+        // }
```
Comment (leiysky) [inline src/query/sql/src/planner/optimizer/rule/rewrite/rule_eliminate_eval_scalar.rs:82]: Remove this?

**Review by leiysky — APPROVED** [review]
Rest LGTM

File: src/query/sql/src/planner/optimizer/rule/rewrite/rule_eliminate_eval_scalar.rs (hunk 1 above)
Comment (xudong963) [reply src/query/sql/src/planner/optimizer/rule/rewrite/rule_eliminate_eval_scalar.rs:82]: Ok

File: src/query/sql/src/planner/optimizer/rule/rewrite/rule_eliminate_eval_scalar.rs [hunk 2]
```diff
@@ -65,24 +63,23 @@ impl Rule for RuleEliminateEvalScalar {
...
-            if new_items.is_empty() {
-                state.add_result(s_expr.child(0)?.clone());
-            } else {
-                state.add_result(SExpr::create_unary(
-                    EvalScalar { items: new_items }.into(),
-                    s_expr.child(0)?.clone(),
-                ));
-            }
-        }
+        // if eval_scalar.items.iter().any(|item| matches!(&item.scalar, ScalarExpr::BoundColumnRef(column) if column.column.index == item.index)) {
```
Comment (BohuTANG) [inline src/query/sql/src/planner/optimizer/rule/rewrite/rule_eliminate_eval_scalar.rs:68]: This code no longer used?

Comment (xudong963) [reply src/query/sql/src/planner/optimizer/rule/rewrite/rule_eliminate_eval_scalar.rs]: Yes

**Pull request merged**